sycl/source/detail/sampler_impl.cpp

```diff
@@ -46,34 +46,40 @@ RT::PiSampler sampler_impl::getOrCreateSampler(const context &Context) {
   if (m_contextToSampler[Context])
     return m_contextToSampler[Context];
 
-#if CL_TARGET_OPENCL_VERSION > 120
-  const cl_sampler_properties sprops[] = {
-      CL_SAMPLER_NORMALIZED_COORDS,
-      static_cast<cl_sampler_properties>(m_CoordNormMode),
-      CL_SAMPLER_ADDRESSING_MODE,
-      static_cast<cl_sampler_properties>(m_AddrMode),
-      CL_SAMPLER_FILTER_MODE,
-      static_cast<cl_sampler_properties>(m_FiltMode),
-      0};
+  // OpenCL spec guaranties that version is in following format:
+  // OpenCL<space><major_version.minor_version><space><platform-specific information>
+  // Let's get these numbers
+  std::string RawOpenCLVersion = Context.get_platform().get_info<sycl::info::platform::version>();
+  std::string OpenCLVersion = split_string(RawOpenCLVersion, ' ')[1];
 
-  PI_CALL((m_contextToSampler[Context] = RT::piSamplerCreate(
-      getSyclObjImpl(Context)->getHandleRef(), sprops, &errcode_ret),
-      errcode_ret));
+  if (OpenCLVersion >= "2.0") {
+    const cl_sampler_properties sprops[] = {
+        CL_SAMPLER_NORMALIZED_COORDS,
+        static_cast<cl_sampler_properties>(m_CoordNormMode),
+        CL_SAMPLER_ADDRESSING_MODE,
+        static_cast<cl_sampler_properties>(m_AddrMode),
+        CL_SAMPLER_FILTER_MODE,
+        static_cast<cl_sampler_properties>(m_FiltMode),
+        0};
 
-  if (errcode_ret == PI_INVALID_OPERATION)
-      throw feature_not_supported("Images are not supported by this device.");
-#else
-  // TODO: do we really need this old interface into PI and here?
-  cl_int cl_errcode_ret;
-  m_contextToSampler[Context] =
-      clCreateSampler(getSyclObjImpl(Context)->getHandleRef(),
-                      static_cast<cl_bool>(m_CoordNormMode),
-                      static_cast<cl_addressing_mode>(m_AddrMode),
-                      static_cast<cl_filter_mode>(m_FiltMode), &cl_errcode_ret);
-  if (cl_errcode_ret == CL_INVALID_OPERATION)
-      throw feature_not_supported("Images are not supported by this device.");
-  PI_CHECK(cl_errcode_ret);
-#endif
+    PI_CALL((m_contextToSampler[Context] = RT::piSamplerCreate(
+        getSyclObjImpl(Context)->getHandleRef(), sprops, &errcode_ret),
+        errcode_ret));
+
+    if (errcode_ret == PI_INVALID_OPERATION)
+        throw feature_not_supported("Images are not supported by this device.");
+  } else {
+    // TODO: do we really need this old interface into PI and here?
+    cl_int cl_errcode_ret;
+    m_contextToSampler[Context] = pi::pi_cast<RT::PiSampler>(
+        clCreateSampler(pi::pi_cast<cl_context>(getSyclObjImpl(Context)->getHandleRef()),
```

Comment:
@smaslov-intel, Is it OK that I call OpenCL 1.0 API like that? In current implementation PI always calls OpenCL 2.0 API so I have to call OpenCL 1.0 API directly. But the SYCL compiler is supposed to work with all target device runtimes regardless their version. I guess it's up to PI to decide which API it should call (I suggest to always call 1.0 API).

Reply:
We should NOT call OpenCL directly ever, except where running OpenCL interop (and even that better go through PI). In this case (where different OpenCL versions have different interfaces) the  PI should be a unified interface that different plugins would be able to implement. To be more specific, if we really need to support OpenCL 1.0, then this if/else should go to the implementation of the piSamplerCreate in OpenCL plugin for PI (now in pi_opencl.cpp), and SYCL RT would just call piSamplerCreate always. OpenCL 2.0 version had already been chosen as an interface, and I don't see a reason to change that now.

Reply:
I think a good existing example of this is perhaps OpenCL PI's piPlatformsGet (correct me if i am wrong though): https://github.com/intel/llvm/blob/sycl/sycl/source/detail/pi_opencl.cpp#L81 

I also plan to pull request in something similar for piQueueCreate in the near future.

Reply:
Yes, exactly. Thanks for pointing this out. @KarachunIvan, please rework the sampler creation accordingly.